### PR TITLE
Doc comments: Handle CDATA correctly and avoid normalizing code elements

### LINKED ISSUE #22431: Metadata-as-source does not display summary text inside CDATA
**Version Used**: 
VS Community 2017 v15.3.5.

**Steps to Reproduce**:
1. Create new console project
2. Add NuGet reference to Microsoft.CodeAnalysis
3. Type BinaryOperatorKind, add using and press F12

**Expected Behavior**:
```
        // Summary:
        //     Represents the >> operator.
        RightShift = 9,
```

**Actual Behavior**:
```
        // Summary:
        //     Represents the operator.
        RightShift = 9,
```

The [source code](https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/Operations/BinaryOperatorKind.cs) contains several CDATA sections, but they are not displayed in VS:

image

### LINKED ISSUE #37503: Quick Info for <code> should not normalize whitespace
This is a follow-up to #35667.

See https://github.com/dotnet/roslyn/pull/35667#issuecomment-507758934.

🔗 Also filed as [AB#1322022](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1322022)

### Sample code

```csharp
/// <summary>
/// <code>
/// code line 1
/// code line 2
/// </code>
/// </summary>
private int value;
```

### Expected result

Quick Info for `value` renders the summary code as a two-line code block with code font.

### Actual result

Quick Info for `value` renders the summary code on one line with code font.

### LINKED ISSUE #53135: Quick Info for <code> XML doc tag should not remove CDATA block
**Quick Info strips out CDATA block completely.**

The &lt;code&gt;XML doc elements often contain a CDATA block so it's not necessary to escape special XML characters, such as &lt;, &gt; and &amp;. However, these blocks are completely ignored in Quick Info.

**Code example:**

``` C#
/// <summary>
/// The code example:
/// <code><![CDATA[
/// if (i < 5 && i > 0)
/// {
///     var list = new List<string>();
/// }
/// ]]>
/// </code>
/// This was the example.
/// </summary>
public void Method1()

```

**Actual result**

obrázok

**Expected result**
Quick info shows the contents of CDATA.

This is related to #37503

### LINKED ISSUE #53782: Quick info in IntelliSense strips whitespaces from <code> blocks
**Version Used**: Visual Studio 16.11 Preview 1 (using the .NET 6 SDK Preview 4)

**Steps to Reproduce**:

1. Use this code snippet:
```csharp
class Foo
{
    /// <summary>
    /// This is a test:
    /// <code>
    /// while (Foo())
    /// {
    ///     Bar();
    /// }
    /// 
    /// Baz();
    /// </code>
    /// </summary>
    void Bar()
    {
    }
}
```
2. Hover on `Bar`
3. The following appears in the IntelliSense quick info tooltip:

image

**Expected Behavior**:
The code inside that \<code> block should preserve the original formatting/spacing/newlines, etc.
Interestingly enough, this works fine when viewing the same XML docs through the .NET API browser.

**Actual Behavior**:
All whitespace characters are lost and the code is just weirdly formatted on a single line.

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #53782
Fixes #53135
Fixes #37503
Fixes #22431

Before:

image

---

After:

image

### CODE REVIEW

File: src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs [hunk 1]
```diff
@@ -102,10 +103,12 @@ public void AppendSingleSpace()
             public void AppendString(string s)
             {
                 EmitPendingChars();
-
-                Builder.Add(new TaggedText(TextTags.Text, s, Style, NavigationTarget.target, NavigationTarget.hint));
+                Builder.Add(new TaggedText(TextTags.Text, NormalizeLineEndings(s), Style, NavigationTarget.target, NavigationTarget.hint));
 
                 _anyNonWhitespaceSinceLastPara = true;
+
+                static string NormalizeLineEndings(string input) => Regex.Replace(input, "(?<!\r)\n", "\r\n");
```
Comment (Youssef1313) [inline src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs:110]: I'd certainly prefer a better solution here. The issue here is that `XText.Value` produces a string with Unix-style line endings (`\n`), causing the end result to have mixed line-endings 😕

File: src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs [hunk 2]
```diff
@@ -169,11 +169,11 @@ private void AddDocumentationContent(ISymbol symbol)
 
                 _documentationMap.Add(
                     SymbolDescriptionGroups.Documentation,
-                    symbol.GetDocumentationParts(_semanticModel, _position, formatter, CancellationToken).ToImmutableArray());
+                    symbol.GetDocumentationParts(_semanticModel, _position, formatter, CancellationToken));
 
                 _documentationMap.Add(
                     SymbolDescriptionGroups.RemarksDocumentation,
-                    symbol.GetRemarksDocumentationParts(_semanticModel, _position, formatter, CancellationToken).ToImmutableArray());
+                    symbol.GetRemarksDocumentationParts(_semanticModel, _position, formatter, CancellationToken));
```
Comment (Youssef1313) [inline src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs:176]: This only avoids creating a new array since `GetDocumentationParts` already returns an immutable array.

File: src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs:110]: this warrants a comment explaning what the regex means and waht this is here for.

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by sharwell** [comment]
Ohhhh snap 🎉

File: src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs [hunk 3]
```diff
@@ -7265,5 +7265,53 @@ void M2()
...
+                Documentation(@"summary for interface IGoo
+
+
+List<string> y = null;
+"));
+        }
+
+        [WorkItem(37503, "https://github.com/dotnet/roslyn/issues/37503")]
+        [Fact, Trait(Traits.Feature, Traits.Features.QuickInfo)]
+        public async Task DoNotNormalizeWhitespaceForCode()
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs:7294]: 💡 Can you add both of these test cases to a new test class **IntellisenseQuickInfoBuilderTests_Code.cs**, which is modeled after `IntellisenseQuickInfoBuilderTests_Styles`.

File: src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs [hunk 4]
```diff
@@ -102,10 +103,15 @@ public void AppendSingleSpace()
...
-                Builder.Add(new TaggedText(TextTags.Text, s, Style, NavigationTarget.target, NavigationTarget.hint));
+                Builder.Add(new TaggedText(TextTags.Text, NormalizeLineEndings(s), Style, NavigationTarget.target, NavigationTarget.hint));
 
                 _anyNonWhitespaceSinceLastPara = true;
+
+                // This regex replaces `\n`s that are not preceded by `\r` with `\r\n`.
+                // This is because XText.Value returns a string contaning `\n` as the line endings, causing
+                // the end result to have mixed line-endings. So normalize everything to `\r\n`.
+                static string NormalizeLineEndings(string input) => Regex.Replace(input, "(?<!\r)\n", "\r\n");
+
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs:114]:
```suggestion
```

File: src/EditorFeatures/Test/DocCommentFormatting/DocCommentFormattingTests.cs [hunk 5]
```diff
@@ -50,7 +50,7 @@ public void ExampleAndCodeTags()
             results in <c>p</c>'s having the value (2,8).
             </example>";
 
-            var expected = "This method changes the point's location by the given x- and y-offsets. For example:\r\n\r\nPoint p = new Point(3,5); p.Translate(-1,3);\r\n\r\nresults in p's having the value (2,8).";
+            var expected = "This method changes the point's location by the given x- and y-offsets. For example:\r\n\r\n\r\n            Point p = new Point(3,5);\r\n            p.Translate(-1,3);\r\n            \r\n\r\nresults in p's having the value (2,8).";
```
Comment (sharwell) [inline src/EditorFeatures/Test/DocCommentFormatting/DocCommentFormattingTests.cs:53]: 📝 This is one of the cases that made me uncomfortable, but I'll reserve judgement until I see the new builder tests.

**Review by sharwell — CHANGES_REQUESTED** [review]

File: src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs (hunk 2 above)
Comment (sharwell) [reply src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs:176]: `ToImmutableArray()` won't clone the array if the input is already an immutable array, but this does avoid boxing the `IEnumerable<T>` for the call and is cleaner so 👍

File: src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs (hunk 1 above)
Comment (sharwell) [reply src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs]: I normally use `input.Replace("\r\n", "\n").Replace("\n", Environment.NewLine)`, but XML is required to normalize line endings to `\n` so I'm guessing you can just use `input.Replace("\n", "\r\n")` and be done.

File: src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs (hunk 2 above)
Comment (Youssef1313) [reply src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs:176]: > ToImmutableArray() won't clone the array if the input is already an immutable array,

@sharwell Yeah that seem the case:

https://github.com/dotnet/runtime/blob/59d9d0d16858635a8c228be2ab7ad8882b0d1c71/src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray.cs#L372-L380

So the docs of CA2009 are inaccurate/incorrect?

https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2009

> Performance issue: Unnecessary creation of a duplicate immutable collection. The original collection was already immutable and can be used directly.

File: src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs (hunk 1 above)
Comment (Youssef1313) [reply src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs]: @sharwell I'm a little bit worried if AppendString can be called with a string already containing `\r\n`, then doing `Replace("\n", "\r\n")` will duplicate `\r`.

Is it's guaranteed that `XText.Value` will **always** be normalized to `\n`?

File: src/EditorFeatures/Test/DocCommentFormatting/DocCommentFormattingTests.cs (hunk 5 above)
Comment (Youssef1313) [reply src/EditorFeatures/Test/DocCommentFormatting/DocCommentFormattingTests.cs:53]: @sharwell I'm expecting you'll still be uncomfortable after seeing the new tests 😄 
Does it make sense to trim line-endings from code elements?

Comment (sharwell) [reply src/EditorFeatures/Test/DocCommentFormatting/DocCommentFormattingTests.cs:53]: I'm going to look at screenshots before requesting changes to the actual normalization approach.

File: src/EditorFeatures/Test2/IntelliSense/IntellisenseQuickInfoBuilderTests_Code.vb [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+                            New ClassifiedTextRun(ClassificationTypeNames.Punctuation, "."),
+                            New ClassifiedTextRun(ClassificationTypeNames.MethodName, "MyMethod", navigationAction:=Sub() Return, "void MyClass.MyMethod()"),
+                            New ClassifiedTextRun(ClassificationTypeNames.Punctuation, "("),
+                            New ClassifiedTextRun(ClassificationTypeNames.Punctuation, ")"))),
+                   New ClassifiedTextElement(
+                       New ClassifiedTextRun(ClassificationTypeNames.Text, "Normalize this, but"),
+                        New ClassifiedTextRun(ClassificationTypeNames.WhiteSpace, " "),
+                        New ClassifiedTextRun(ClassificationTypeNames.Text, "Not        this", ClassifiedTextRunStyle.UseClassificationFont))),
+                New ClassifiedTextElement(
+                    New ClassifiedTextRun(ClassificationTypeNames.Text, "\r\nline 1\r\nline     2\r\n", ClassifiedTextRunStyle.UseClassificationFont)),
```
Comment (sharwell) [inline src/EditorFeatures/Test2/IntelliSense/IntellisenseQuickInfoBuilderTests_Code.vb:105]: 💭  I would expect this to be rendered like `<br/>`

Example input:

https://github.com/dotnet/roslyn/blob/3e1cf13ade005df2497a4bf789381cb99024f5f8/src/EditorFeatures/Test2/IntelliSense/IntellisenseQuickInfoBuilderTests.vb#L326-L327

Example output:

https://github.com/dotnet/roslyn/blob/3e1cf13ade005df2497a4bf789381cb99024f5f8/src/EditorFeatures/Test2/IntelliSense/IntellisenseQuickInfoBuilderTests.vb#L371-L376

Comment (sharwell) [reply src/EditorFeatures/Test2/IntelliSense/IntellisenseQuickInfoBuilderTests_Code.vb:105]: I'll have to run the code to see if it looks the same with the current form.

Comment (Youssef1313) [reply src/EditorFeatures/Test2/IntelliSense/IntellisenseQuickInfoBuilderTests_Code.vb:105]: I'll add screenshots shortly.

Comment (Youssef1313) [reply src/EditorFeatures/Test2/IntelliSense/IntellisenseQuickInfoBuilderTests_Code.vb:105]: @sharwell Screenshot added.

**Review by Youssef1313 — COMMENTED** [review]
While playing around this, I found that indentation of code isn't preserved. This is due to:

https://github.com/dotnet/roslyn/blob/1cca63b5d8ea170f8d8e88e1574aa3ebe354c23b/src/Workspaces/Core/Portable/Shared/Utilities/DocumentationComment.cs#L182

I'll open an issue to address this as a follow-up. Not sure if we should simply drop calling `TrimEachLine` or if it should understand the structure of XML and avoids trimming things inside `<code>` but trim other lines.

File: src/EditorFeatures/Test2/IntelliSense/IntellisenseQuickInfoBuilderTests_Code.vb [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+                            New ClassifiedTextRun(ClassificationTypeNames.Punctuation, "."),
+                            New ClassifiedTextRun(ClassificationTypeNames.MethodName, "MyMethod", navigationAction:=Sub() Return, "void MyClass.MyMethod()"),
+                            New ClassifiedTextRun(ClassificationTypeNames.Punctuation, "("),
+                            New ClassifiedTextRun(ClassificationTypeNames.Punctuation, ")"))),
+                   New ClassifiedTextElement(
+                       New ClassifiedTextRun(ClassificationTypeNames.Text, "Normalize this, but"),
+                        New ClassifiedTextRun(ClassificationTypeNames.WhiteSpace, " "),
+                        New ClassifiedTextRun(ClassificationTypeNames.Text, "Not        this", ClassifiedTextRunStyle.UseClassificationFont))),
+                New ClassifiedTextElement(
+                    New ClassifiedTextRun(ClassificationTypeNames.Text, $"{vbCrLf}line 1{vbCrLf}line     2{vbCrLf}", ClassifiedTextRunStyle.UseClassificationFont)),
```
Comment (sharwell) [inline src/EditorFeatures/Test2/IntelliSense/IntellisenseQuickInfoBuilderTests_Code.vb:105]: 💡 The leading and trailing newlines should be dropped.

```suggestion
                    New ClassifiedTextRun(ClassificationTypeNames.Text, $"line 1{vbCrLf}line     2", ClassifiedTextRunStyle.UseClassificationFont)),
```

Comment (Youssef1313) [reply src/EditorFeatures/Test2/IntelliSense/IntellisenseQuickInfoBuilderTests_Code.vb]: @sharwell Sorry for the delay. Fixed.

File: src/EditorFeatures/Test2/IntelliSense/IntellisenseQuickInfoBuilderTests_Code.vb [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+        <WpfFact, Trait(Traits.Feature, Traits.Features.QuickInfo)>
+        Public Async Function QuickInfoForXmlCodeElement() As Task
+            Dim workspace =
+                <Workspace>
+                    <Project Language="C#" CommonReferences="true">
+                        <Document>
+                            class MyClass {
+                                /// &lt;summary&gt;
+                                /// Normalize    this, but &lt;c&gt;Not        this&lt;/c&gt;
```
[the remark this answers is not in the document]
Comment (Youssef1313) [reply src/EditorFeatures/Test2/IntelliSense/IntellisenseQuickInfoBuilderTests_Code.vb:68]: @sharwell This is still code, so I'm expecting it to preserve spaces. But no idea what tools like docfx do.

**Review by sharwell — APPROVED** [review]

File: src/Features/Core/Portable/Common/TaggedTextStyle.cs [hunk 9]
```diff
@@ -13,12 +11,14 @@ internal enum TaggedTextStyle
...
+        Emphasis = 1 << 1,
 
-        Emphasis = 0x2,
+        Underline = 1 << 2,
 
-        Underline = 0x4,
+        Code = 1 << 3,
 
-        Code = 0x8,
+        InlineCode = 1 << 4,
```
Comment (Youssef1313) [inline src/Features/Core/Portable/Common/TaggedTextStyle.cs:22]: I added a new TaggedTextStyle to differentiate `<c>` and `<code>`.

Currently `Code` value is `1000` and `InlineCode` is `10000`. Should `InlineCode` be `1001` or whatever value to share the same most significant bit? (i.e, indicate that InlineCode is a subset of Code)?

This will allow the following check: `(style & TaggedTextStyle.Code) == TaggedTextStyle.Code` to pass whether `style` is `Code` or `InlineCode`.

File: src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs [hunk 10]
```diff
@@ -695,7 +695,7 @@ static string GetStyledText(TaggedText taggedText, bool isInCodeBlock)
                     TaggedTextStyle.Strong => $"**{text}**",
                     TaggedTextStyle.Emphasis => $"_{text}_",
                     TaggedTextStyle.Underline => $"<u>{text}</u>",
-                    TaggedTextStyle.Code => $"`{text}`",
+                    TaggedTextStyle.Code or TaggedTextStyle.InlineCode => $"`{text}`",
```
Comment (Youssef1313) [inline src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs:698]: Few lines above this, there is:

```csharp
                if (!isInCodeBlock)
                    text = text.Replace(" ", "&nbsp;");
```

This doesn't feel correct. See the following test:

https://github.com/dibarbet/roslyn/blob/0ef88b0a686892546b57e721ba65738872bf928d/src/Features/LanguageServer/ProtocolUnitTests/Hover/HoverTests.cs#L195

@sharwell Do you agree here? Do you want me to fix LSP in this PR or make that a separate PR?

cc: @dibarbet

File: src/Features/Core/Portable/Common/TaggedTextStyle.cs (hunk 9 above)
Comment (sharwell) [reply src/Features/Core/Portable/Common/TaggedTextStyle.cs:22]: The current approach seems fine to me.

File: src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs (hunk 10 above)
Comment (sharwell) [reply src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs:698]: Separate is fine.

File: src/Features/Core/Portable/Common/TaggedTextStyle.cs (hunk 9 above)
Comment (sharwell) [reply src/Features/Core/Portable/Common/TaggedTextStyle.cs:22]: On second thought, it seems like the new addition should be:

```
PreserveWhitespace = 1 << 4,
```

**Review by sharwell — CHANGES_REQUESTED** [review]
Let's change `<code>` to use `TaggedTextStyle.Code | TaggedTextStyle.PreserveWhitespace` so the whole enumeration continues to work as individual flags.

**Review by sharwell — APPROVED** [review]

**Comment by Youssef1313** [comment]
@sharwell Anything else needed here?

**Comment by sharwell** [comment]
@Youssef1313 We'd like to rebase this onto **release/dev16.11**. Are you able to do this or would you like us to go ahead and do it?

**Comment by Youssef1313** [comment]
> @Youssef1313 We'd like to rebase this onto **release/dev16.11**. Are you able to do this or would you like us to go ahead and do it?

Done.

**Comment by sharwell** [comment]
@jinujoseph for approval to merge

File: src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs [hunk 11]
```diff
@@ -102,10 +103,14 @@ public void AppendSingleSpace()
...
-
-                Builder.Add(new TaggedText(TextTags.Text, s, Style, NavigationTarget.target, NavigationTarget.hint));
+                Builder.Add(new TaggedText(TextTags.Text, NormalizeLineEndings(s), Style, NavigationTarget.target, NavigationTarget.hint));
 
                 _anyNonWhitespaceSinceLastPara = true;
+
+                // XText.Value returns a string with `\n` as the line endings, causing
+                // the end result to have mixed line-endings. So normalize everything to `\r\n`.
+                // https://www.w3.org/TR/xml/#sec-line-ends
+                static string NormalizeLineEndings(string input) => input.Replace("\n", "\r\n");
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs:113]: why do we need this though?  what is wrong with the doc comment having \n in it?

File: src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs [hunk 12]
```diff
@@ -305,8 +310,9 @@ public ImmutableArray<TaggedText> Format(string rawXmlText, ISymbol symbol, Sema
 
         private static void AppendTextFromNode(FormatterState state, XNode node, Compilation compilation)
         {
-            if (node.NodeType == XmlNodeType.Text)
+            if (node.NodeType is XmlNodeType.Text or XmlNodeType.CDATA)
             {
+                // cast is safe since XCData inherits XText
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs:315]: should the above just be `if (node is XText xtext)` ?

File: src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs [hunk 13]
```diff
@@ -550,6 +559,13 @@ private static string TrimCrefPrefix(string value)
         private static void AppendTextFromTextNode(FormatterState state, XText element)
         {
             var rawText = element.Value;
+            if ((state.Style & TaggedTextStyle.PreserveWhitespace) == TaggedTextStyle.PreserveWhitespace)
+            {
+                // Don't normalize code from middle. Only trim leading/trailing new lines.
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs:564]: why trim the leading/trailing?

Comment (sharwell) [reply src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs:564]: ➡️ This implements a fix which a test revealed:
https://github.com/dotnet/roslyn/pull/53140#discussion_r628561538

File: src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs (hunk 12 above)
Comment (sharwell) [reply src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs:315]: ➡️ Both approaches are fine. There isn't really a downside to the current approach so I prefer to not make a change here unless it's being done along with other changes.

File: src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs (hunk 11 above)
Comment (sharwell) [reply src/Features/Core/Portable/DocumentationComments/AbstractDocumentationCommentFormattingService.cs:113]: ➡️ This function ensures that non-normalized XML content uses the same end-of-line character as `s_newlinePart`. It could be cleaned up in the future (e.g. if we wanted to use `Environment.Newline` instead), but the approach in this PR keeps the result consistent with current precedent.

**Comment by Youssef1313** [comment]
@sharwell Can you close the linked issues? Thanks!

**Comment by sharwell** [comment]
@Youssef1313 Already done thanks 🎉🎉
